### PR TITLE
Add new cop Style/NestedFunctionDefinition

### PR DESCRIPTION
**Pull request by ojab**

### CODE REVIEW

File: config/enabled.yml [hunk 1]
```diff
@@ -402,6 +402,10 @@ Style/NegatedWhile:
   StyleGuide: 'https://github.com/bbatsov/ruby-style-guide#until-for-negatives'
   Enabled: true
 
+Style/NestedFunctionDefinition:
```
Comment (jonas054) [inline config/enabled.yml]: Call it `Style/NestedMethodDefinition`.

File: lib/rubocop/cop/style/nested_function_definition.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# encoding: utf-8
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks for nested function definitions.
+      class NestedFunctionDefinition < Cop
```
Comment (jonas054) [inline lib/rubocop/cop/style/nested_function_definition.rb]: Include the `OnMethodDef` mixin to support singleton method definitions as well.

**Comment by bbatsov** [comment]
I'd suggest making this a lint cop. It also needs more specs.

**Comment by bbatsov** [comment]
@ojab Ping

File: config/enabled.yml [hunk 3]
```diff
@@ -402,6 +402,10 @@ Style/NegatedWhile:
...
+  Description: 'Do not use nested function definitions.'
+  Enabled: true
```
Comment (bbatsov) [inline config/enabled.yml]: This should also have a StyleGuide link.

File: lib/rubocop/cop/style/nested_function_definition.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+# encoding: utf-8
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks for nested function definitions.
```
Comment (bbatsov) [inline lib/rubocop/cop/style/nested_function_definition.rb]: I'd expand the comment here with an explanation of the problem and code snippets.

File: lib/rubocop/cop/style/nested_function_definition.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+      class NestedFunctionDefinition < Cop
+        MSG = 'Function declarations must not be nested. ' \
```
Comment (bbatsov) [inline lib/rubocop/cop/style/nested_function_definition.rb]: Function -> Method

**Comment by ojab** [comment]
Sorry for delay, pushed new version with [I hope] all issues fixed.

File: spec/rubocop/cop/lint/nested_method_definition_spec.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+                         'end',
+                         'foo = Foo.new',
+                         'def foo.bar',
+                         '  def baz',
+                         '  end',
+                         'end'])
+    expect(cop.offenses.size).to eq(1)
+  end
+
+  it 'register an offence for a nested method definition inside lambda' do
```
Comment (bbatsov) [inline spec/rubocop/cop/lint/nested_method_definition_spec.rb]: registers

File: spec/rubocop/cop/lint/nested_method_definition_spec.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    inspect_source(cop, ['def foo',
+                         '  bar = -> { def baz; puts; end  }',
+                         'end'])
+    expect(cop.offenses.size).to eq(1)
+  end
+
+  it 'doesn not an offence for a lambda definition inside method' do
```
Comment (bbatsov) [inline spec/rubocop/cop/lint/nested_method_definition_spec.rb]: does not register

File: spec/rubocop/cop/lint/nested_method_definition_spec.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                         'def foo.bar',
+                         '  def baz',
+                         '  end',
+                         'end'])
+    expect(cop.offenses.size).to eq(1)
+  end
+
+  it 'register an offence for a nested method definition inside lambda' do
+    inspect_source(cop, ['def foo',
+                         '  bar = -> { def baz; puts; end  }',
```
Comment (bbatsov) [inline spec/rubocop/cop/lint/nested_method_definition_spec.rb]: There's an extra space here, before the closing `}`.

File: spec/rubocop/cop/lint/nested_method_definition_spec.rb [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+  it 'doesn not an offence for a lambda definition inside method' do
+    inspect_source(cop, ['def foo',
+                         '  bar = -> { puts  }',
+                         '  bar.call',
+                         'end'])
+    expect(cop.offenses.size).to eq(0)
+  end
+
+  it 'doesn not registers an offense for a nested class method definition' do
```
Comment (bbatsov) [inline spec/rubocop/cop/lint/nested_method_definition_spec.rb]: does not register

**Comment by bbatsov** [comment]
This should be accompanied by a changelog entry.

**Comment by ojab** [comment]
And it looks like I've done something that prevent travis-ci/coveralls triggering :/

**Pull request closed without merging**